### PR TITLE
WIP: Add cloudevents endpoints to server

### PR DESCRIPTION
**Pull request by technosophos**
This adds CloudEvents support to the server.

### CODE REVIEW

**Comment by technosophos** [comment]
I have no idea why `dep` reformatted the lock file...

**Comment by radu-matei** [comment]
A newer version of `dep` reformats the lock file this way - the original file was generated with an older version.

File: cmd/gateway/server.go [hunk 1]
```diff
@@ -106,3 +111,53 @@ func azFn(c *gin.Context) {
 	c.JSON(http.StatusOK, gin.H{"message": "OK"})
 	return
 }
+
+// ceFn is a cloud events handler.
+func ceFn(c *gin.Context) {
```
Comment (radu-matei) [inline cmd/gateway/server.go:116]: The first thing we're going to need here is endpoint validation - https://docs.microsoft.com/en-us/azure/event-grid/security-authentication


When creating the event subscription, the first event will be a `Microsoft.EventGrid.SubscriptionValidationEvent` (which does not appear to be CloudSpec compliant?)

```
[
    {
        "id": "be0d5d8f-585b-4ace-861f-fc4beaaf2477",
        "topic": "/subscriptions/<subscription-id>/resourceGroups/eventgrid-cloudevents/providers/microsoft.storage/storageaccounts/evcloudeventsstorage",
        "subject": "",
        "data": {
            "validationCode": "validation-code",
            "validationUrl": "https://rp-northeurope.eventgrid.azure.net/eventsubscriptions/local-ngrok/validate?id=validation-code&t=2018-05-25T11:41:08.7774182Z&apiVersion=2018-05-01-preview&token=token"
        },
        "eventType": "Microsoft.EventGrid.SubscriptionValidationEvent",
        "eventTime": "2018-05-25T11:41:08.8086405Z",
        "metadataVersion": "1",
        "dataVersion": "2"
    }
]
```

It expects a response of type:

```
{
    "validationResponse": "validation-code"
}
```

Comment (radu-matei) [reply cmd/gateway/server.go:116]: Yes, it appears that the validation event does not follow the CloudEvents spec, but the EventGrid default schema. So when we receive a request, we need to decode it into an EventGrid type - if that fails, try to decode it as CloudSpec event.

**Review by radu-matei — CHANGES_REQUESTED** [review]

**Comment by radu-matei** [comment]
Once we validate the endpoint, we can start receiving events from Event Grid - here's how a `blob deleted` event looks like:

```
{
  "eventType": "Microsoft.Storage.BlobDeleted",
  "eventTypeVersion": "",
  "cloudEventsVersion": "0.1",
  "source": "/subscriptions/<subscription-id>/resourceGroups/eventgrid-cloudevents/providers/Microsoft.Storage/storageAccounts/evcloudeventsstorage#blobServices/default/containers/testing-container/blobs/README.md",
  "eventID": "a90602e3-201e-0051-261e-f4bd9106719c",
  "eventTime": "2018-05-25T11:50:39.0091025Z",
  "data": {
    "api": "DeleteBlob",
    "requestId": "<request-id>",
    "eTag": "<e-tag>",
    "contentType": "application/octet-stream",
    "contentLength": 141,
    "blobType": "BlockBlob",
    "url": "https://evcloudeventsstorage.blob.core.windows.net/testing-container/README.md",
    "sequencer": "000000000000000000000000000001B100000000028f3beb",
    "storageDiagnostics": {
      "batchId": "<batch-id>"
    }
  }
}
```

**Pull request merged**